src/main/utility/mod.rs

```diff
@@ -362,70 +383,112 @@ impl std::fmt::Display for VerifyPluginPathError {
+        // Linux allows recursion here. It has a depth limit of 4, but trying to
+        // check and precisely match that doesn't seem worthwhile.
+        log::debug!("{path:?} has interpreter {interp:?}; checking");
+        verify_plugin_path(interp)
+            .map_err(|e| VerifyPluginPathError::IncompatibleInterpreter(Box::new(e)))
+    } else {
+        Err(VerifyPluginPathError::UnknownFileType)
+    }
+}
+
+/// Check that the plugin path is executable under Shadow.
+// Memoization wrapper around `verify_plugin_path_internal`
+// TODO: maybe move this cache into `sim_config.rs`? This seems slightly more
+// possible to go stale for paths exec'd by managed code.
+pub fn verify_plugin_path(path: impl AsRef<std::path::Path>) -> Result<(), VerifyPluginPathError> {
+    let path = path.as_ref();
 
     // a cache so we don't check the same path multiple times (assuming the user doesn't move any
     // binaries while shadow is running)
-    // TODO: maybe move this into `sim_config.rs`? This seems slightly more
-    // possible to go stale for paths exec'd by managed code.
-    static CHECKED_DYNAMIC_BINS: Lazy<RwLock<HashSet<PathBuf>>> =
-        Lazy::new(|| RwLock::new(HashSet::new()));
+    static CHECKED_BINS: Lazy<RwLock<HashSet<PathBuf>>> = Lazy::new(|| RwLock::new(HashSet::new()));
 
-    let is_known_dynamic = CHECKED_DYNAMIC_BINS.read().unwrap().contains(path);
+    if CHECKED_BINS.read().unwrap().contains(path) {
+        return Ok(());
+    }
 
-    // check if the binary is dynamically linked
-    if !is_known_dynamic {
-        let ld_path = "/lib64/ld-linux-x86-64.so.2";
-        let ld_output = std::process::Command::new(ld_path)
-            .arg("--verify")
-            .arg(path)
-            .output()
-            .expect("Unable to run '{ld_path}'");
-
-        if ld_output.status.success() {
-            CHECKED_DYNAMIC_BINS
-                .write()
-                .unwrap()
-                .insert(path.to_path_buf());
-        } else {
-            log::debug!("ld stderr: {:?}", ld_output.stderr);
-            // technically ld-linux could return errors for other reasons, but this is the most
-            // likely reason given that we already checked that the file exists
-            return Err(VerifyPluginPathError::NotDynamicallyLinkedElf);
-        }
+    let res = verify_plugin_path_internal(path);
+    if res.is_ok() {
+        CHECKED_BINS.write().unwrap().insert(path.to_path_buf());
     }
+    res
+}
 
-    Ok(())
+fn get_interpreter(header: &[u8]) -> Option<&Path> {
+    // Verify and strip "shebang"
+    let mut header = header.strip_prefix(b"#!")?;
+    // Skip any spaces. (Other whitespace isn't skipped AFAIK).
+    while header.first() == Some(&b' ') {
+        header = &header[1..];
+    }
+    // The path is the next contiguous set of non-space or newline characters.
+    let interp_path = match header.splitn(2, |b| b == &b' ' || b == &b'\n').next() {
+        Some(i) => i,
+        None => header,
```

Comment:
Changing this to return `None` in this case instead of returning a *prefix* of the interpreter path.